src/python_testing/TC_SU_2_2.py

```diff
@@ -0,0 +1,1251 @@
+                if not final_idle_seen:  # log only once
+                    final_idle_seen = True
+                    state_sequence.append(val)
+                    logger.info(f'{step_number}: 1.2.1 - OTA UpdateState sequence complete, final state is Idle')
+            # Return True only when Idle is reached
+            return final_idle_seen
+
+        # Create matcher object from UpdateState
+        matcher_update_state_obj = AttributeMatcher.from_callable(
+            description="Validate OTA UpdateState transitions: Downloading > Applying > Idle",
+            matcher=matcher_update_state
+        )
+
+        progress_recorded = set()
+        progress_values = []
+        progress_seen = False
+        final_null_seen = False
+
+        def matcher_progress(report):
+            """
+            Step #1.2.2 matcher function to track OTA progress values.
+            Tracks state transitions: non-null progress values range 1-99 > None.
+            Records each unique value once and expects a final None after progress has been observed.
+            """
+            nonlocal progress_values, progress_seen, final_null_seen, progress_recorded
+            val = getattr(report.value, "value", report.value)  # unwrap Nullable if needed
+
+            if val == NullValue:
+                val = None
+
+            # Track progress numbers 1-99
+            if val is not None and isinstance(val, int) and 1 <= val <= 99:
+                progress_seen = True
+                if val not in progress_recorded:
+                    progress_recorded.add(val)
+                    progress_values.append(val)
+                    logger.info(f'{step_number}: 1.2.2 - UpdateStateProgress recorded: {val}')
+                return False
+
+            # Track final None after some progress
+            if val is None and progress_seen:
+                final_null_seen = True
+                progress_values.append(val)
+                logger.info(f'{step_number} 1.2.2 - OTA UpdateStateProgress sequence complete, final state is None: {progress_values}')
+                return final_null_seen
+
+            return False
+
+        # Create matcher object for UpdateStateProgress
+        matcher_progress_obj = AttributeMatcher.from_callable(
+            description="Validate OTA UpdateStateProgress transitions: non-null values rage 1-99 and final None",
+            matcher=matcher_progress
+        )
+
+        # ------------------------------------------------------------------------------------
+        # [STEP_1]: Step #1.3 - Start tasks to track OTA attributes:
+        # UpdateState and UpdateStateProgress (updateAvailable sequence) with validations
+        # ------------------------------------------------------------------------------------
+        try:
+            # Wait until the final state (Idle) is reached or timeout (5 min)
+
+            thread_state = threading.Thread(
```

Comment:
With this threading implementation, are you sure that exceptions thrown in the thread will be propagated up to the test layer?

This actually seems like overkill - there's a queue in place here. You can just do these in the main thread. If the second completes before the first, you should still have all the expected reports in the queue. You don't need to catch them real time.